Index: src/main/java/org/codehaus/mojo/versions/CompareDependenciesMojo.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/main/java/org/codehaus/mojo/versions/CompareDependenciesMojo.java	(revision 19377)
+++ src/main/java/org/codehaus/mojo/versions/CompareDependenciesMojo.java	(revision 19378)
@@ -315,7 +315,7 @@
             Dependency dependency = remoteDependencies.get( id );
             if ( dependency == null )
             {
-                getLog().warn( "Could not update ${" + property.getName() + "}: unknown dependency " + id );
+                getLog().info( "Not updating ${" + property.getName() + "}: no info for " + id );
                 return null;
             }
             else
\ No newline at end of file
